9567: correct solve_linear return values and capture in solve

smichr opened the pull request:
closes #9567

asmeurer commented:
Does this fix the original issue in #9567 as well?

asmeurer commented:
I don't understand what solve_linear does, but if it fixes the issue then +1.

smichr commented:
I showed on the issue page that the reported issue returns correctly. 

As for solve_linear, it is trying to solve for a linear variable and also report (early in the solve process) if there is no solution possible (like when a  variable appears in an expression but actually cancels after simplification).

I can't believe we didn't see this error earlier.

smichr commented on sympy/solvers/solvers.py:
```diff
@@ -1956,12 +1957,17 @@ def solve_linear(lhs, rhs=0, symbols=[], exclude=[]):
-                        if not d.has(xi) or not (d/xi).has(xi):
+                        if xi not in dfree:
+                            return xi, vi
+                        if xi not in (d/xi).free_symbols:
```
think more about this...seems like this is equivalent to `xi not in dn.free_symbols` which is already the condition of this `if-block`. Perhaps both of these tests are not necessary and `return xi, vi` is all that is needed. Then factor_terms would not be needed.

[ ] check to see what (if anything) gets by these two tests.

asmeurer commented:
I guess what I don't understand is why it sometimes returns a solution and sometimes returns the numerator and denominator.

smichr committed:
factor free_symbol calculation out of comprehension

smichr committed:
use single element rather than single element list

smichr committed:
modifications of solve_linear

- clarify inline note regarding solution from solve_linear
- simplify the testing of the return value from solve_linear
- clarify solve_linear docstring
- modify solve_linear
 o make sure that if there is no symbol to solve for that 0, 0 is returned
 o modify routine in keeping with original but relax condition that the symbol
   returned must not appear in the denominator or must appear in a linear
   fashion. The routine will return a solution if the numerator is linear,
   regardless of what the denominator.

smichr committed:
edit minsolve_linear_system docstring

smichr commented:
> why it sometimes returns a solution and sometimes returns the numerator and denominator

it should probably have been a private function of solve; I think the original idea was to not throw away the extracted numerator which is needed for the more extended work of solve. Since the function was returning 2 values when it found a solution it made sense to return the denominator with the numerator, too, so 2 values would be returned when there wasn't a linear solution.
